### PR TITLE
collision_monitor: dynamic polygon and source enable/disable

### PR DESCRIPTION
**Pull request by tonynajjar**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | https://github.com/ros-planning/navigation2/issues/3824 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Own gazebo simulation |

---

## Description of contribution in a few bullet points

* Added dynamic enabling/disabling of collision monitor polygons and sources
* Added test

## Description of documentation updates required from your changes

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_collision_monitor/src/polygon.cpp [hunk 1]
```diff
@@ -487,6 +500,26 @@ void Polygon::updatePolygon(geometry_msgs::msg::PolygonStamped::ConstSharedPtr m
...
+{
+  rcl_interfaces::msg::SetParametersResult result;
+
+  for (auto parameter : parameters) {
+    const auto & param_type = parameter.get_type();
+    const auto & param_name = parameter.get_name();
+
+    if (param_type == rcl_interfaces::msg::ParameterType::PARAMETER_BOOL) {
+      if (param_name == polygon_name_ + "." + "enabled") {
+        enabled_ = parameter.as_bool();
```
Comment (tonynajjar) [inline nav2_collision_monitor/src/polygon.cpp:515]: setter function better?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Minor changes but LGTM. Some test would be nice though

File: nav2_collision_monitor/src/polygon.cpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_collision_monitor/src/polygon.cpp:515]: Nope, this is fine!

File: nav2_collision_monitor/src/collision_monitor_node.cpp [hunk 2]
```diff
@@ -382,6 +382,10 @@ void CollisionMonitor::process(const Velocity & cmd_vel_in)
   std::shared_ptr<Polygon> action_polygon;
 
   for (std::shared_ptr<Polygon> polygon : polygons_) {
+    if (!polygon->getEnabled()) {
+      // If the polygon is disabled, skip it
```
[the remark this answers is not in the document]
Comment (tonynajjar) [reply nav2_collision_monitor/src/collision_monitor_node.cpp:386]: Done. One more question, do you think I should also skip publishing the polygon in `publishPolygons` if it's disabled?

**Comment by tonynajjar** [comment]
Also, since collision monitor and detector use the same Polygon, detector polygons will also have this new parameter. We could implement the same logic in collision_detector but I don't see much benefit to it except consistency/parallelism between the two nodes. What do you think?

**Comment by tonynajjar** [comment]
> Doesn't hurt anyone and really easy.

One drawback I see to disabling a polygon in collision_detector is that it would mess up the indexing of polygons in `CollisionDetectorState` msg; if some node subscribing to this topic has saved at initialization that a polygon is at a certain index, then disabling a polygon would cause this node to get the wrong index in `detections` (or even access an index out of range).

The reason why I said I see no value in disabling a polygon in collision detector is because if you don't want the status of a certain polygon, you just don't use it.

If you still think that we should add the functionality and it's safe to expect that the user should always check if `CollisionDetectorState` msg has changed, then I'm fine with implementing it

**Comment by tonynajjar** [comment]
@SteveMacenski in my last commit I also added the dynamic enabling/disabling of a source

**Review by tonynajjar — COMMENTED** [review]
I have a general question about dynamic parameter setting: from the Nav2 codebase I see that you prefer having these dynamicParametersCallback in which we set internal member variables with the new parameters.

Another way would be to directly read the parameter from the parameter server, roughly:


```
    auto node = polygon->node_.lock();
    if (!node->get_parameter(polygon->getName() + ".enabled").as_bool()) {
      continue;
    }
```

The benefit here is less boilerplate code. What's the con, is it slower? Or why do you prefer the first option?

**Comment by SteveMacenski** [comment]
> if some node subscribing to this topic has saved at initialization that a polygon is at a certain index

That's really not good practice and I'd put that into the category of "not our problem". Also, aren't only the polygons in collision published? If so, then there's no order to be assumed because there's no consistent polygon collision. 

> If you still think that we should add the functionality

I do

> dynamicParametersCallback in which we set internal member variables with the new parameters

Because updating a parameter may require reinitializing other areas that parameter is being used in or you could be triggering an action mid-parameter update if trying to update many - so then the state of that action is not fully defined. Further, design wise, its not wise to inject ROS code into your deep algorithm logic, so getting the ROS params and converting to internal types lets you keep ROS out of the library cores + gives you control to explicitly accept or reject parameter updates if they're bogus or invalid, which happens

File: nav2_collision_monitor/src/source.cpp [hunk 3]
```diff
@@ -37,10 +37,23 @@ Source::Source(
   transform_tolerance_(transform_tolerance), source_timeout_(source_timeout),
   base_shift_correction_(base_shift_correction)
 {
+  RCLCPP_INFO(logger_, "[%s]: Creating Source", source_name_.c_str());
```
Comment (SteveMacenski) [inline nav2_collision_monitor/src/source.cpp:40]: Remove here and destructor

**Comment by SteveMacenski** [comment]
@AlexeyMerzlyakov please review

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
@tonynajjar just needs migration guide entry for navigation.ros.org + configuration guide for the new parameters

**Comment by tonynajjar** [comment]
https://github.com/ros-planning/navigation.ros.org/pull/471

**Comment by SteveMacenski** [comment]
Great! Thanks! This is nice to have!

**Pull request merged**